File: lib/core/aegis/routing/router_service.dart
import '../domain/execution_status.dart';
import '../domain/provider_request.dart';
import '../domain/provider_response.dart';
import '../domain/routing_info.dart';
import 'provider_registry.dart';
import 'task_classifier.dart';

typedef RoutingProgressCallback = void Function(RoutingInfo info);

class RouterService {
  RouterService({
    TaskClassifier? classifier,
    ProviderRegistry? registry,
  })  : _classifier = classifier ?? const TaskClassifier(),
        _registry = registry ?? ProviderRegistry();

  final TaskClassifier _classifier;
  final ProviderRegistry _registry;

  ProviderRegistry get registry => _registry;

  Future<ProviderResponse> route(
    ProviderRequest request, {
    RoutingProgressCallback? onProgress,
  }) async {
    onProgress?.call(
      const RoutingInfo(
        taskType: '—',
        selectedProvider: '—',
        executionStatus: 'Classifying',
      ),
    );

    final classification = _classifier.classify(request);
    final provider = _registry.resolve(classification.taskType);

    onProgress?.call(
      RoutingInfo(
        taskType: classification.taskType.displayLabel,
        selectedProvider: provider.name,
        executionStatus: ExecutionStatus.routing.label,
        confidence: classification.confidence,
      ),
    );

    onProgress?.call(
      RoutingInfo(
        taskType: classification.taskType.displayLabel,
        selectedProvider: provider.name,
        executionStatus: ExecutionStatus.executing.label,
        confidence: classification.confidence,
      ),
    );

    try {
      final response = await provider.execute(
        request,
        classification: classification,
      );

      onProgress?.call(
        RoutingInfo(
          taskType: response.taskType.displayLabel,
          selectedProvider: response.providerName,
          executionStatus: ExecutionStatus.completed.label,
          confidence: classification.confidence,
        ),
      );

      return response;
    } catch (e) {
      onProgress?.call(
        RoutingInfo(
          taskType: classification.taskType.displayLabel,
          selectedProvider: provider.name,
          executionStatus: ExecutionStatus.failed.label,
          confidence: classification.confidence,
        ),
      );

      return ProviderResponse(
        taskType: classification.taskType,
        providerId: provider.id,
        providerName: provider.name,
        executionStatus: ExecutionStatus.failed,
        content: 'Execution failed: $e',
        errorMessage: e.toString(),
      );
    }
  }

  List<Map<String, dynamic>> listProviders() {
    return _registry.all
        .map(
          (p) => {
            'id': p.id,
            'name': p.name,
            'taskType': p.taskType.value,
            'isMock': p.isMock,
          },
        )
        .toList();
  }
}

File: backend/src/core/aegis/router-service.js
import { ExecutionStatus, TaskType } from './task-types.js';
import { TaskClassifier } from './task-classifier.js';
import { ProviderRegistry } from './provider-registry.js';

const TASK_LABELS = {
  [TaskType.TEXT]: 'Text',
  [TaskType.CODE]: 'Code',
  [TaskType.IMAGE]: 'Image',
  [TaskType.VIDEO]: 'Video',
  [TaskType.RESEARCH]: 'Research',
};

export class RouterService {
  constructor({ classifier, registry } = {}) {
    this._classifier = classifier ?? new TaskClassifier();
    this._registry = registry ?? new ProviderRegistry();
  }

  get registry() {
    return this._registry;
  }

  listProviders() {
    return this._registry.list().map((p) => ({
      id: p.id,
      name: p.name,
      taskType: p.taskType,
      isMock: p.isMock,
    }));
  }

  async route(request, onProgress) {
    onProgress?.({
      taskType: '—',
      selectedProvider: '—',
      executionStatus: ExecutionStatus.CLASSIFYING,
    });

    const classification = this._classifier.classify(request);
    const provider = this._registry.resolve(classification.taskType);
    const taskLabel = TASK_LABELS[classification.taskType] ?? classification.taskType;

    onProgress?.({
      taskType: taskLabel,
      selectedProvider: provider.name,
      executionStatus: ExecutionStatus.ROUTING,
      confidence: classification.confidence,
    });

    onProgress?.({
      taskType: taskLabel,
      selectedProvider: provider.name,
      executionStatus: ExecutionStatus.EXECUTING,
      confidence: classification.confidence,
    });

    try {
      const result = await provider.execute(request, classification);

      onProgress?.({
        taskType: TASK_LABELS[result.taskType] ?? result.taskType,
        selectedProvider: result.providerName,
        executionStatus: ExecutionStatus.COMPLETED,
        confidence: classification.confidence,
      });

      return {
        ...result,
        routing: {
          taskType: TASK_LABELS[result.taskType] ?? result.taskType,
          selectedProvider: result.providerName,
          executionStatus: ExecutionStatus.COMPLETED,
          confidence: classification.confidence,
        },
        classification,
      };
    } catch (err) {
      onProgress?.({
        taskType: taskLabel,
        selectedProvider: provider.name,
        executionStatus: ExecutionStatus.FAILED,
        confidence: classification.confidence,
      });

      return {
        taskType: classification.taskType,
        providerId: provider.id,
        providerName: provider.name,
        executionStatus: ExecutionStatus.FAILED,
        content: `Execution failed: ${err.message}`,
        errorMessage: err.message,
        routing: {
          taskType: taskLabel,
          selectedProvider: provider.name,
          executionStatus: ExecutionStatus.FAILED,
          confidence: classification.confidence,
        },
        classification,
      };
    }
  }
}
